Client/Frontend/UserContent/UserScripts/DeAMP.js

```diff
@@ -0,0 +1,63 @@
+// Copyright 2022 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+"use strict";
+
+(_ => {
+  const W = window
+  const D = W.document
+
+  let timesToCheck = 20
+  let intervalId
+
+  const checkIfShouldStopChecking = _ => {
+    timesToCheck -= 1
+    if (timesToCheck === 0) {
+      W.clearInterval(intervalId)
+    }
+  }
+
+  const checkForAmp = _ => {
+    const htmlElm = document.documentElement
+    const headElm = document.head
+
+    if (!headElm && !htmlElm) {
+      // checked too early and page structure not available.
+      checkIfShouldStopChecking()
+      return
+    }
+
+    if (!htmlElm.hasAttribute('amp') && !htmlElm.hasAttribute('⚡')) {
+      // We know this isn't an amp document, and no point of checking further
+      W.clearInterval(intervalId)
+      return
+    }
+
+    const canonicalLinkElm = D.querySelector('head > link[rel="canonical"][href^="http"]')
+    if (!canonicalLinkElm === null) {
+      // didn't find a link elm.
+      checkIfShouldStopChecking()
+      return
+    }
+
+    const targetHref = canonicalLinkElm.getAttribute('href')
+    try {
+      const destUrl = new URL(targetHref)
+      if (W.location.href == destUrl.href) {
+        // Avoid an infinite loading loop in some cases
+        W.clearInterval(intervalId)
+        return
+      }
+      W.location.replace(destUrl.href)
```

Comment:
Can we replace this call with something like:

```javascript
window.open('...', '_blank')
window.close()
```

Can we also verify that the protocol is either `http` or `https`?

`if (destUrl.href.protocol !== 'http' || destUrl.href.protocol !== 'https'` should return eagerly.

This will safeguard from malicious links running in the current google amp context.

Reply:
```
window.open('...', '_blank')
window.close()
```

I dont think we should do this.  This will open a new window and close the current one, which means you'll loose back-forward, etc.

@thypon is there a security reason to prefer the above?  `W.location = <url>` is (as far as i know) the/a common way to redirect the page.

> Can we also verify that the protocol is either http or https?

This seems like a great idea

Reply:
@thypon `window.location.replace` doesn't have a target parameter. But also, any attempts to redirect to a `javascript` URL will execute the Javascript immediately and not redirect. If somehow the URL is decoded to be `javascript`, it is caught on the redirect on the iOS side via: https://github.com/brave/brave-ios/blob/development/Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift#L139

which is `if url.isBookmarklet { ..cancel.. }` to cancel the navigation.